### ISSUE TITLE
Unstable results on every new process when I restarted the server

### ISSUE THREAD

**Issue by tonyabracadabra** [microsoft/LightGBM#5628]
### Lightgbm version:
'3.3.3'

### Platform
Reproducible on both MacOS and Linux

### Code

```python3

import lightgbm as lgb

def generate_trees(X, y):
    params = {
        'objective': 'binary',
         'deterministic': True,
         'force_row_wise': True,
         'num_threads': 8,
         'learning_rate': 0.05,
         'max_depth': 3,
         'tree_learner': 'data',
         'boosting_type': 'gbdt',
         'feature_fraction': 0.7,
         'bagging_fraction': 0.8,
         'bagging_freq': 5,
         'verbose': -1,
         'seed': 8,
     }
     random.seed(8)
     np.random.seed(8)

     dataset = lgb.Dataset(X, y)
     bst = lgb.train(params, dataset, 10, keep_training_booster=True)
     trees = bst.dump_model()["tree_info"]

    return trees

def compare_trees(trees_one, trees_second) -> bool:
    ...

trees_list = []
for i in range(100):
    trees = generate_trees(X, y)
    trees_list.append(trees)

# this works just fine because all the running is on a same process
for i in range(len(trees_list)):
    for j in range(len(trees_list)):
        assert compare_trees(trees_list[i], trees_list[j])

# but as I open a new process, running the algorithm again with the new data, we will get this
assert compare_trees(trees_new_on_new_process, trees_on_prev_process) is False
```

I was testing it out on my local macbook M2 machine with a flask server.
- The code does give me stable result when I run the algorithm for multiple times without restarting the process.
- Every time when I reboot the server, it always give me a new randomized version of the code.
- I've tested out every possible instruction in https://lightgbm.readthedocs.io/en/latest/Parameters.html
- I've even tried setting the random seed of python / numpy, but it doesn't work
- I've also disabled parallelism by setting `num_threads` to 1, but it doesn't help

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

Can you please try to provide a [minimal, reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)? I don't understand what you mean by the following:

* "run the algorithm in a row"
* "new randomized version of the code"
* "setting the random seed of Python"

Please also provide the details that the issue template asked for, including:

* what version of LightGBM?
* what commands did you run to install it?
* what operating system are you on?

-----

I see you've also double-posted this to Stack Overflow ([link](https://stackoverflow.com/questions/74759691/unstable-lightgbm-results-on-every-new-process-when-i-restarted-the-server)). Please do not do that. 

We monitor the `[lightgbm`] tag on Stack Overflow just like we monitor this issues page.... it's possible that I could have been responding to you on Stack Overflow at the same time someone else was responding here, which would have been a waste of maintainers' very limited time. Double-posting can also make it less likely that others facing the same issue will be able to find the answer from search engines.
